web/server.js

```diff
@@ -190,36 +203,49 @@ module.exports = function(options) {
             method: function(request, reply) {
               oauthDb.getClient(request.query.client_id, reply);
             }
-          },
-          {
-            assign: 'auth_code',
-            method: function(request, reply) {
-              var scopes = request.query.scopes;
-              var expiresAt = new Date(Date.now() + 60 * 1000).toISOString();
-
-              scopes = scopes.split(' ');
-
-              oauthDb.generateAuthCode(
-                request.pre.client.client_id,
-                request.pre.user.id,
-                scopes,
-                expiresAt,
-                reply
-              );
-            }
           }
         ]
       },
       handler: function(request, reply) {
+        var responseType = request.query.response_type;
+        var client = request.pre.client;
+
+        if (
+          client.allowed_responses.indexOf(responseType) === -1
+        ) {
+          return reply(Boom.forbidden('Response type forbidden: ' + responseType));
+        }
+
+        var scopes = request.query.scopes;
         var state = request.query.state;
-        var redirectObj = url.parse(request.pre.client.redirect_uri, true);
-        redirectObj.search = null;
-        redirectObj.query.code = request.pre.auth_code;
-        redirectObj.query.state = state;
-        redirectObj.query.client_id = request.query.client_id;
-        var redirectUri = url.format(redirectObj);
-
-        reply.redirect(redirectUri);
+        var expiresAt = new Date(Date.now() + 60 * 1000).toISOString();
+        var redirectObj = url.parse(client.redirect_uri, true);
+
+        scopes = scopes.split(' ');
+
+        oauthDb.generateAuthCode(
```

Comment:
Lets move this generateAuthCode function back into a pre method, and put the generateToken function in a pre method as well. We can have them conditionally run based on the response_type, and also return an error if it's not allowed. Then the handler can just be output formatting!